Redis custom chef recipe

radamanthus opened the pull request:
## Description of your patch

Updates the Redis recipe to use the installer from redis.io. This lets us support newer Redis versions, even ones still in alpha.

## Recommended Release Notes

Updates the Redis recipe to use the installer from redis.io.

## Estimated risk

Low

## Components involved

redis custom chef recipe

## Description of testing done

See QA instructions

## QA Instructions

Run the test cases in examples/redis/cookbooks/custom-redis/README.md

radamanthus commented:
Avoid running the installer if possible. Put the binary in the recipe as a `cookbook_file`.

radamanthus commented:
Adding the binaries to the cookbook will bloat the cookbook. Let's just install from the Gentoo package (default) and add the option to install by downloading the installer.

aespinosa commented on examples/redis/cookbooks/custom-redis/README.md:
````diff
@@ -0,0 +1,128 @@
+* Ensure that these lines are not commented out:
+
+```
+  redis['utility_name'] = 'redis'
````
if `is_redis_instance` is used, why is this still needed?

radamanthus replied:
To be consistent with how `utility_name` is used in the other recipes, e.g. thinking_sphinx:
https://github.com/engineyard/ey-cookbooks-stable-v5/blob/ac50d45d8281bb7284a49eda452bf3cf73cba878/cookbooks/thinking_sphinx_3/attributes/default.rb#L24

The recipe is easier to use if you only have to set utility_name, instead of modifying the is_redis_instance logic.

aespinosa replied:
ok that got me confused.  i thought we're using just `is_XXX_instance`  in other cookbooks. https://github.com/engineyard/ey-cookbooks-stable-v5/blob/next-release/cookbooks/delayed_job4/recipes/default.rb#L6

aespinosa commented on examples/redis/cookbooks/custom-redis/attributes/default.rb:
```diff
@@ -0,0 +1,97 @@
+default['redis'].tap do |redis|
+  # Set to true if you want to install from source
+  # Installing from the Gentoo package in the portage tree is faster,
```
Also note that installing from packages is the better 'devops' approach.

radamanthus replied:
I updated the comment to say that installing from packages is the recommended approach.

aespinosa commented on cookbooks/redis/recipes/install_from_source.rb:
```diff
@@ -0,0 +1,29 @@
+  command "tar zxf redis-#{redis_version}.tar.gz && sync"
+end
+
+execute "Remove old redis-source" do
```
Is this needed because redis tarball has the same directory? It's probably bettero to extract to `/opt/redis-<version>` . This might causes headaches debugging things in production.

radamanthus replied:
I updated install_from_source to etxract to /opt/redis-version.

radamanthus commented:
I updated the PR. Summary of changes:

- Mention in attributes/default.rb comment that installing from package is the recommended approach
- Use action :install instead of action :upgrade
- Extract to /opt/redis-version instead of /opt/redis-source - this is easier to debug because it avoids the possibility of multiple archives being extracted to the same installer directory 
- Removed masked packages from the list of available version

Re: offering pre-built versions of redis outside the portage tree - it will bloat main chef if we include the binaries. If we put the binaries elsewhere, then we'll end up maintaining a binary repository similar to the portage tree - why not just update the portage tree then? 

The approach I ended up with is a middle ground:
- install from packages in the portage tree
- install from source, only if what you need is not in the portage tree and you can't wait for the portage tree update
- work with Engineering to update the portage tree packages

rbankston commented:
Looks good to me
